### PR TITLE
Validate user-supplied missing values on unmapped fields

### PR DESCRIPTION
**Pull request by not-napoleon**
This PR adds the ability for Aggregation factories to have finer grained control over how missing values are mapped to `ValuesSource`s.  Technically speaking, this addresses a minor bug in the terms aggregation, but more importantly it resolves a design constraint for Range field aggregations.  I opened the PR against master however because it doesn't depend on any of the range field aggregation work and touches shared infrastructure for aggregations, so getting it in master sooner will avoid merge conflicts down the road.

I've added a bunch of tests, and done my best to preserve the existing behavior wherever possible.  The only noticeable change should be that if you pass something exceedingly unusual as a missing value for an unmapped field to the terms agg, it will no longer silently convert it to a string and continue.  Instead, it will throw an `AggregationExecutionException` same as if you'd had that value in a mapped field.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregatorFactory.java [hunk 1]
```diff
@@ -43,13 +43,17 @@ public ValuesSourceAggregatorFactory(String name, ValuesSourceConfig<VS> config,
...
             List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) throws IOException {
-        VS vs = config.toValuesSource(context.getQueryShardContext());
+        VS vs = config.toValuesSource(context.getQueryShardContext(), this::resolveMissingAny);
         if (vs == null) {
             return createUnmapped(parent, pipelineAggregators, metaData);
         }
         return doCreateInternal(vs, parent, collectsFromSingleBucket, pipelineAggregators, metaData);
     }
 
+    protected ValuesSource resolveMissingAny(Object missing) {
```
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregatorFactory.java]: Let's add some javadocs on this method so it's clear what this is for (and why a subclass might want to override it)

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorTests.java [hunk 2]
```diff
@@ -188,6 +198,80 @@ public void testMissing() throws Exception {
...
+            }
+
+            HistogramAggregationBuilder aggBuilder = new HistogramAggregationBuilder("my_agg")
+                .field("field")
+                .interval(5)
+                .missing("🍌🍌🍌");
+            MappedFieldType type = null;
+            try (IndexReader reader = w.getReader()) {
+                IndexSearcher searcher = new IndexSearcher(reader);
+                expectedException.expect(IllegalArgumentException.class);
```
Comment (polyfractal) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorTests.java]: Unless there is a strong reason to use the junit `expectedException` Rule thing, let's convert these to `expectThrows()`?  Makes it a little more clear what is supposed to throw an exception, and prevents other accidental exceptions from being caught/processed.

File: server/src/test/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorTests.java [hunk 3]
```diff
@@ -107,8 +108,61 @@ public void testQueryFiltersAll() throws IOException {
...
+        }, null);
+    }
+
+    public void testUnmappedMissingGeoPoint() throws IOException {
+        CardinalityAggregationBuilder aggregationBuilder = new CardinalityAggregationBuilder("name", null)
+            .field("number").missing(new GeoPoint(42.39561, -71.13051));
+
+        testCase(aggregationBuilder, new MatchAllDocsQuery(), iw -> {
+            iw.addDocument(singleton(new NumericDocValuesField("unrelatedField", 7)));
+                iw.addDocument(singleton(new NumericDocValuesField("unrelatedField", 8)));
```
Comment (polyfractal) [inline server/src/test/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorTests.java]: Nit: spacing :)

File: server/src/test/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorTests.java [hunk 4]
```diff
@@ -117,13 +171,8 @@ private void testCase(Query query, CheckedConsumer<RandomIndexWriter, IOExceptio
...
         IndexSearcher indexSearcher = newSearcher(indexReader, true, true);
 
-        CardinalityAggregationBuilder aggregationBuilder = new CardinalityAggregationBuilder(
-                "_name", ValueType.NUMERIC).field("number");
-        MappedFieldType fieldType = new NumberFieldMapper.NumberFieldType(
-                NumberFieldMapper.NumberType.LONG);
-        fieldType.setName("number");
         CardinalityAggregator aggregator = createAggregator(aggregationBuilder, indexSearcher,
-            fieldType);
+            fieldType1);
```
Comment (polyfractal) [inline server/src/test/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorTests.java]: Does this need to be `fieldType1`?

**Review by polyfractal — CHANGES_REQUESTED** [review]
Core changes look good, left a few minor comments/questions.

> The only noticeable change should be that if you pass something exceedingly unusual as a missing value for an unmapped field to the terms agg, it will no longer silently convert it to a string and continue. Instead, it will throw an AggregationExecutionException same as if you'd had that value in a mapped field.

Do we have a good feeling for how "breaking" this is?  E.g. is there a case where someone was using this behavior (accidentally or otherwise) and now we'll throw an exception? I'm not really sure what you could pass in the REST API that would be non-string/non-numeric... an object I guess? Or is it only an issue for java users that can send in objects which have a valid `toString()`?

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregatorFactory.java (hunk 1 above)
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregatorFactory.java]: Done.

Comment (polyfractal) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregatorFactory.java]: :tada:  awesome thanks

**Review by polyfractal — APPROVED** [review]
LGTM, although I am curious about the bwc question in my last comment.  E.g. we should probably backport this to 7.x, but if it's a "real" breaking change it might be tricky/impossible.

**Comment by not-napoleon** [comment]
@polyfractal Ok, as per conversation, I updated the PR so there's zero behavior change, and this just adds hooks and tests.

**Comment by polyfractal** [comment]
:+1: when CI is happy it LGTM!

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Pull request merged**